### PR TITLE
Custom Date Range: Allow to Include Only Start or End Date

### LINKED ISSUE #37: Allow Custom Date Ranges to Include Only Start or End Date
Use case: "show me all results after 2020-01-01." Basically, don't force the user to select both.

Currently, the custom date range aggregation requires that both be set.

### PR DESCRIPTION
**Pull request by renatonascalves**
fixes #37

### CODE REVIEW

File: lib/aggregations/class-custom-date-range.php [hunk 1]
```diff
@@ -72,16 +72,50 @@ function( $query_value ) {
...
+	/**
+	 * Get the default date range.
+	 *
+	 * @param string $range The range to get the default date for.
+	 * @return string
+	 */
+	private function get_default_date_range( string $range ): string {
+		switch ( $range ) {
+			case 'from':
+				$datetime = gmdate( 'Y-m-d', 0 ); // Defaults to 1970-01-01.
```
Comment (renatonascalves) [inline lib/aggregations/class-custom-date-range.php:94]: This works but I'm unsure if `gmdate` is the right function here. 🤔 

Does it matter it is in GMT since it will use `wp_timezone` later to create and format the date?! I guess not.

**Review by mboynes — CHANGES_REQUESTED** [review]
This can be _much_ simpler. ES doesn't require both `from` and `to` to be present. All that needs to happen is `DSL::build_range()` needs to be modified to allow either argument to be `null`, and if one is null, its key (`from` or `to`, as appropriate) is omitted.

**Comment by renatonascalves** [comment]
@mboynes Tks for the feedback. Let me check that approach.

**Comment by renatonascalves** [comment]
Updated and tested as per code review. 💯

File: lib/aggregations/class-custom-date-range.php [hunk 2]
```diff
@@ -95,9 +95,12 @@ private function get_date_range( string $from, string $to ) : array {
...
 			$to_datetime   = DateTime::createFromFormat( DATE_W3C, $to );
-			return $from_datetime && $to_datetime
-				? $this->dsl->build_range( $from_datetime, $to_datetime )
-				: [];
+
+			if ( empty( $from_datetime ) && empty( $to_datetime ) ) {
+				return [];
+			}
+
+			return $this->dsl->build_range( $from_datetime, $to_datetime );
```
Comment (mboynes) [inline lib/aggregations/class-custom-date-range.php:103]: this seems like an unnecessary change -- or am I missing something?

File: lib/class-dsl.php [hunk 3]
```diff
@@ -158,16 +158,23 @@ public static function all_terms( string $field, array $values ): array {
 	 * Given DateTime objects for the from and to components of a range query,
 	 * builds the from/to components of the range query and returns it.
 	 *
-	 * @param DateTime $from The start date.
-	 * @param DateTime $to   The end date.
+	 * @param string|DateTime $from The start date.
+	 * @param string|DateTime $to   The end date.
```
Comment (mboynes) [inline lib/class-dsl.php:162]: `DateTime::createFromFormat()` either returns a `DateTime` or `false`, it would never produce a string. Further, we'd definitely never want it to be a string.

I'd suggest that this be `DateTime|null`, the types in the function signature be `?DateTime`, and in `Custom_Date_Range::get_date_range()`, that call each argument as e.g. `$from_datetime ?: null`.

File: lib/aggregations/class-custom-date-range.php (hunk 2 above)
Comment (renatonascalves) [reply lib/aggregations/class-custom-date-range.php]: No, you are right. Removing it. It is not needed.

File: lib/class-dsl.php [hunk 4]
```diff
@@ -158,16 +158,23 @@ public static function all_terms( string $field, array $values ): array {
...
-			'to'   => $to->format( 'Y-m-d H:i:s' ),
-		];
+	public function build_range( $from, $to ): array {
+		$range = [];
+
+		if ( ! empty( $from ) && $from instanceof DateTime ) {
+			$range['from'] = $from->format( 'Y-m-d H:i:s' );
+		}
+
+		if ( ! empty( $to ) && $to instanceof DateTime ) {
```
Comment (mboynes) [inline lib/class-dsl.php:173]: These empty checks aren't bringing anything to the table

```suggestion
		if ( $from instanceof DateTime ) {
			$range['from'] = $from->format( 'Y-m-d H:i:s' );
		}

		if ( $to instanceof DateTime ) {
```

**Review by mboynes — CHANGES_REQUESTED** [review]

**Comment by renatonascalves** [comment]
CR changes applied. ✅

**Review by mboynes — APPROVED** [review]
💯
